### DISCUSSION TITLE
Adding time-integral of a variable in PeleC

### DISCUSSION THREAD

**Discussion by RSuryaNarayan** [AMReX-Combustion/PeleC#487] [Q&A] [answered]
Hello, 

I would like to know if there's a way to add the time integral of a variable in PeleC. Say ```phi(x,y,t)``` is a variable in the plot file, I would like to find out ```integral phi(x,y,t) dt```. If we discretize this expression, its simply a weighted sum of ```phi``` with the ```dt```s at the various levels.

I approached this by trying to define a new variable in ```Derive.cpp``` but couldn't find the necessary API to reference ```dt``` for multiple levels? Am I approaching this the right way or is there a shorter way to do this? 

Thanks a lot !

**Comment by marchdf** [comment, marked as answer]
Again, apologies for missing this. `parent->dtLevel(level)` is the api to get the `dt` at a given `level`. Hopefully this helps!

**Reply by RSuryaNarayan** [reply]
@marchdf Thanks a ton for the answer! I will try it out and let you know.

**Reply by RSuryaNarayan** [reply]
@marchdf I was trying a work around to the integral variable. However this new method requires the previous timestep data in the derived variable functions (like ```derpress``` etc.) Is there a way I can access that fab too? currently I can only see current state data in the ```dat``` fab. 

Thanks a lot

**Reply by marchdf** [reply]
You can see how we explicitly call the derived functions here: https://github.com/AMReX-Combustion/PeleC/blob/development/Source/PeleC.cpp#L1364. You could try passing in something that comes from the old state instead of the new state. Depending on when you are doing that in the code, it might work.

**Reply by RSuryaNarayan** [reply]
Thanks for this. However, this is the place where you call them. I assume their definitions are all present in ```Derive.cpp``` and ```Derive.H``` and ideally, I'd like to add something like ```old_fab``` to the argument list there. Something like this:

```
pc_newvariable(
  const amrex::Box& bx,
  amrex::FArrayBox& derfab, 
 amrex::FArrayBox& old_datfab, // This one should hold the state data for the previous timestep
  int dcomp,
  int ncomp,
  const amrex::FArrayBox& datfab,
  const amrex::Geometry& geomdata,
  amrex::Real time,
  const int* bcrec,
  int level)
```

Could I know if this would be possible by simply changing their definitions in the ```Derive`` files?

**Reply by marchdf** [reply]
Oh I see. hmmmm it feels like you could do that and then grab the old state. `amrex::MultiFab& S_old = get_old_data(State_Type);`

**Reply by RSuryaNarayan** [reply]
Yes exactly. I tried this precise line inside those derived variable function definitions but ran into scope issues....
```
/home/suryanarayan/Pele/PeleC_sootfoil/Source/Derive.cpp: In function ‘void pc_dersootfoil(const amrex::Box&, amrex::FArrayBox&, int, int, const amrex::FArrayBox&, const amrex::Geometry&, amrex::Real, const int*, int)’:
/home/suryanarayan/Pele/PeleC_sootfoil/Source/Derive.cpp:319:28: error: ‘get_old_data’ was not declared in this scope
  319 |   amrex::MultiFab& S_old = get_old_data(State_Type);
      |                            ^~~~~~~~~~~~
make: *** [/home/suryanarayan/Pele/PeleC_sootfoil/Submodules/AMReX/Tools/GNUMake/Make.rules:199: tmp_build_dir/o/2d.gnu.MPI.EXE/Derive.o] Error 1
make: *** Waiting for unfinished jobs....
```
Maybe I should do this in ```PeleC.cpp```

**Reply by marchdf** [reply]
yes I would do it in PeleC.cpp and then pass it in.

**Reply by RSuryaNarayan** [reply]
something that works inside ```Derive.cpp```: 

```
 PeleC pc_obj;
  amrex::MultiFab& S_old = pc_obj.get_old_data(State_Type);
```
This should hopefully grab the old state!

**Reply by RSuryaNarayan** [reply]
here's what I tried finally but run into segfaults: 
```
void
pc_dernewvar(
  const amrex::Box& bx,
  amrex::FArrayBox& derfab,
  int dcomp,
  int ncomp,
  const amrex::FArrayBox& datfab,
  const amrex::Geometry& geomdata,
  amrex::Real time,
  const int* bcrec,
  int level)
{
  auto const dat = datfab.const_array();
  auto pfab = derfab.array();
  PeleC pc_obj;
  amrex::MultiFab& S_old = pc_obj.get_old_data(State_Type);
  for (amrex::MFIter mfi(S_old, amrex::TilingIfNotGPU()); mfi.isValid();
       ++mfi) {
    const amrex::Box& box = mfi.tilebox();
    auto old_fab = S_old.array(mfi);
  amrex::ParallelFor(bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
    const amrex::Real rho = dat(i, j, k, URHO);
    const amrex::Real rho_old = old_fab(i, j, k, URHO);
    const amrex::Real rhoInv = 1.0 / rho;
    const amrex::Real rhoInv_old = 1.0 / rho_old;
    amrex::Real T = dat(i, j, k, UTEMP);
    amrex::Real T_old = old_fab(i, j, k, UTEMP);
    // amrex::Real e = dat(i, j, k, UEINT) * rhoInv;
    amrex::Real p, p_old;
    amrex::Real massfrac[NUM_SPECIES];
    amrex::Real massfrac_old[NUM_SPECIES];
    for (int n = 0; n < NUM_SPECIES; ++n) {
      massfrac[n] = dat(i, j, k, UFS + n) * rhoInv;
      massfrac_old[n] = old_fab(i, j, k, UFS + n) * rhoInv_old;      
    }
    auto eos = pele::physics::PhysicsType::eos();
    eos.RTY2P(rho, T, massfrac, p);
    eos.RTY2P(rho_old, T_old, massfrac_old, p_old);
    pfab(i, j, k) = p_old;
  });
  }
}
```
This is by grabbing the old state inside the function. I am not sure how this can be done outside in ```PeleC.cpp``` as it only seems to call these functions rather than define it...their definitions are all in ```Derive.cpp```.

**Reply by marchdf** [reply]
yeah that won't work. Can't you define the function that takes in a Fab of the old data and gets called in PeleC? So you move this `amrex::MultiFab& S_old = pc_obj.get_old_data(State_Type); [...] auto old_fab = S_old.array(mfi); ` to the call site and pass in `old_fab` into the derive call?

**Reply by RSuryaNarayan** [reply]
I will do that but the only place this function seems to be called is at the place where tagging happens. To simply add it as a new plotfile variable would I need to call it somewhere else again? Thanks once again!

**Reply by RSuryaNarayan** [reply]
@marchdf I also tried experimenting with the functions ```post_step``` and looks like adding a new "state" variable to the list will do my job. Could I also know if simply adding a new macro to ```IndexDefines.H``` and modifying the NVARS suffice for this?
Thanks

**Reply by marchdf** [reply]
Yup that could work too. You could augment the state to hold some n-1 states too.

**Reply by RSuryaNarayan** [reply]
Cool! would I have to increase the primitive vars too by, say, n-1, if I increase the state by n-1?

**Reply by marchdf** [reply]
not if you don't store the primitive vars of n-1

**Reply by RSuryaNarayan** [reply]
@marchdf thanks a ton for the prompt answers here. Finally managed to get what I wanted working! Post_step with a new state variable did the job.

**Reply by marchdf** [reply]
Fantastic!
